src/main/remote-sync/RemoteSyncManager.ts

```diff
@@ -0,0 +1,347 @@
+import Logger from 'electron-log';
+import * as helpers from './helpers';
+import {
+  RemoteSyncStatus,
+  RemoteSyncedFolder,
+  RemoteSyncedFile,
+  SyncConfig,
+  SYNC_OFFSET_MS,
+} from './helpers';
+import { reportError } from '../bug-report/service';
+import { obtainToken } from '../auth/service';
+import { DatabaseAdapter } from '../database/adapters/base';
+
+export class RemoteSyncManager {
+  private foldersSyncStatus: RemoteSyncStatus = 'IDLE';
+  private filesSyncStatus: RemoteSyncStatus = 'IDLE';
+  private status: RemoteSyncStatus = 'IDLE';
+  private onStatusChangeCallback: (newStatus: RemoteSyncStatus) => void =
+    () => {
+      return;
+    };
+  constructor(
+    private db: {
+      files: DatabaseAdapter<RemoteSyncedFile>;
+      folders: DatabaseAdapter<RemoteSyncedFolder>;
+    },
+    private config: {
+      fetchFilesLimitPerRequest: number;
+      fetchFoldersLimitPerRequest: number;
+      syncFiles: boolean;
+      syncFolders: boolean;
+    }
+  ) {}
+
+  onStatusChange(callback: (newStatus: RemoteSyncStatus) => void) {
+    this.onStatusChangeCallback = callback;
+  }
+  getSyncStatus(): RemoteSyncStatus {
+    return this.status;
+  }
+
+  isSynced() {
+    return this.status === 'SYNCED';
+  }
+
+  /**
+   * Triggers a remote sync so we can populate the localDB, this sync
+   * is global and starts pulling all the files the user has in remote.
+   *
+   * Throws an error if there's a sync in progress for this class instance
+   */
+  async startRemoteSync() {
+    if (this.status === 'SYNCING') {
+      throw new Error(
+        'RemoteSyncManager is already SYNCING, wait until the current sync finishes'
+      );
+    }
+
+    await this.db.files.connect();
+    await this.db.folders.connect();
+
+    Logger.info('Starting RemoteSyncManager');
+    this.changeStatus('SYNCING');
+    try {
+      await Promise.all([
+        this.config.syncFiles
+          ? this.syncRemoteFiles({
+              retry: 1,
+              maxRetries: 3,
+            })
+          : Promise.resolve(),
+        this.config.syncFolders
+          ? this.syncRemoteFolders({
+              retry: 1,
+              maxRetries: 3,
+            })
+          : Promise.resolve(),
+      ]);
+    } catch (error) {
+      this.changeStatus('SYNC_FAILED');
+      reportError(error as Error);
+    }
+  }
+
+  private changeStatus(newStatus: RemoteSyncStatus) {
+    this.status = newStatus;
+    this.onStatusChangeCallback(newStatus);
+  }
+
+  private checkRemoteSyncStatus() {
+    // We only syncing files
+    if (
+      this.config.syncFiles &&
+      !this.config.syncFolders &&
+      this.filesSyncStatus === 'SYNCED'
+    ) {
+      this.changeStatus('SYNCED');
```

Comment:
Wouldn't we need an early return after changing the status? If we are only syncing files but the `foldersSyncStatus` is `SYNC_FAILED` we are going to change it here but will be overridden after `Files OR Folders sync failed, RemoteSync Failed`

Reply:
you are right, modified it